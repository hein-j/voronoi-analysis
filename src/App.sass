$primary-color: pink
$font: 'Roboto', sans-serif
$footer-height: 52px
$cell-border: 1px solid gray
$min-width: 320px

%graphic-container
  width: 100%
  height: 100%
  display: grid
  align-items: center
  justify-items: center

body
  margin: 0
  font-family: $font

div
  //border: 1px solid black
  box-sizing: border-box

button
  cursor: pointer

.centered-container
  display: grid
  grid-template-columns: repeat(2, minmax(0, 1fr))
  grid-template-rows: repeat(2, minmax(0, 1fr))
  width: 100%
  height: 100vh
  min-height: 500px
  padding: 20px
  padding-bottom: $footer-height + 15px
  cursor: zoom-in

.cell
  width: 100%
  height: 100%
  padding: 20px

.cell:nth-child(1), .cell:nth-child(2)
  border-bottom: $cell-border

.cell:nth-child(odd)
  border-right: $cell-border

@media(max-width: 820px)
  .centered-container
    grid-template-columns: minmax(0, 1fr)
    grid-template-rows: repeat(4, auto)
    width: 100%
    min-width: $min-width
    height: max-content
    min-height: max-content

  .centered-container > .cell
    border: none
    border-bottom: $cell-border

  .centered-container > .cell:nth-child(4)
    border: none
