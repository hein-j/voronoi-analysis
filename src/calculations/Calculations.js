import CountUp from 'react-countup';
import './Calculations.sass';

function Calculations (props) {
  const {name, value} = props;

  function countDecimals (value) {
    const arrays = value.toString().split('.');
    if (arrays.length === 1) {
      return 0
    }
    return arrays[1].length;
  }

  const decimals = countDecimals(value);

  return (
    <div className="calculations-container">
      <span className="value-name">{name}</span>
      <CountUp className="value"
        start={0}
        end={value}
        duration={1}
        decimals={decimals}
        />
    </div>
  )
}

export default Calculations;