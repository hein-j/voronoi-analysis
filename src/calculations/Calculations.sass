@use '../App'
  
.calculations-container
  @extend %graphic-container
  align-content: center

.value-name
  font-size: 60px
  display: block
  text-align: center

.value
  font-size: 50px
  text-align: center
  width: 40vw
  display: block
  margin-top: 20px
  overflow: hidden
  text-overflow: ellipsis

#popup .value
  width: 100%
  max-width: max-content
  min-width: 0px
  overflow-wrap: anywhere
  text-align: left
  margin: auto
  margin-top: 20px
  

@media(max-width: 820px)
  .value
    width: 80vw
    min-width: 200px

@media(max-width: 420px)
  .value-name
    font-size: 40px
  
  .value
    font-size: 30px
