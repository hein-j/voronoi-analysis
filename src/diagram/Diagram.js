import * as d3 from "d3";
import {useEffect} from "react";
import './Diagram.sass';

function Diagram (props) {

  const {positions, apices} = props;

  let renderPositions = [...positions];
  let renderMaxes = [apices.max[0], apices.max[1]];

  // shift if minimums fall in the negative
  if (apices.min[0] < 0 || apices.min[1] < 0) {
    let xShift = 0;
    let yShift = 0;
    if (apices.min[0] < 0) {
      xShift = -1 * apices.min[0];
    }
    if (apices.min[1] < 0) {
      yShift = -1 * apices.min[1];
    }
    renderPositions = renderPositions.map((position) => {
      return [
        (position[0] + xShift),
        (position[1] + yShift)
      ]
    });
    renderMaxes = [renderMaxes[0] + xShift, renderMaxes[1] + yShift];
  }


  // find max between x and y maxes
  const renderMax = Math.max(renderMaxes[0], renderMaxes[1]);

  // find factor to enlarge or shrink by
  const width = 1000;
  const height = 1000;
  const factor = width / renderMax;

  renderPositions = renderPositions.map((position) => {
    return [
      position[0] * factor,
      position[1] * factor
    ]
  });

  // have first three positions in beginning of animation
  let tempPositions = renderPositions.slice(0,3);

  useEffect(renderDiagram, [positions, apices]);
  
  function renderDiagram () {
    // big help from https://observablehq.com/@d3/circle-dragging-iii?collection=@d3/d3-delaunayv
  
    document.querySelector(".diagram-container").innerHTML = '';
  
    const svg = d3.select(".diagram-container")
    .append("svg")
      .attr("viewBox", [0, 0, width, height])
      .attr("stroke-width", 2);
    
    let voronoi = d3.Delaunay
      .from(tempPositions)
      .voronoi([0, 0, width, height]);
  
    const circle = svg.append("g")
      .selectAll("circle")
      .data(renderPositions)
      .join("circle")
        .attr("cx", d => d[0])
        .attr("cy", d => d[1])
        .attr("r", 1)
        .attr("fill", "black");
  
    const mesh = svg.append("path")
        .attr("fill", "none")
        .attr("stroke", "black")
        .attr("stroke-width", 1)
        .attr("d", voronoi.render());
  
    function update(tempPositions) {
      const voronoi = d3.Delaunay.from(tempPositions).voronoi([0, 0, width, height]);
      mesh.attr("d", voronoi.render());
    }
  
    for (let i = 0 ; i < renderPositions.length; i++) {
      // skip if initial first three positions
      if (i < 3) {
        continue;
      }
      setTimeout(()=> {
        tempPositions.push(renderPositions[i]);
        update(tempPositions)
      },i * 2)
    }
  }

  return (
    <div className="diagram-container"></div>
  )
}

export default Diagram;