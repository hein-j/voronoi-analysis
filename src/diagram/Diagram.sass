@use '../App'
  
.diagram-container
  @extend %graphic-container
  
  svg
    max-width: 100%
    max-height: 100%
