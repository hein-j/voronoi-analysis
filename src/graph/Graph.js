import * as d3 from "d3";
import {useEffect} from "react";


function Graph (props) {

  const {areas} = props;

  // create intervals along x-axis

  const maxArea = areas.reduce((max, curr) => {
    return Math.max(max, curr)
    }, 0)
  const numberOfAreas = areas.length;
  const barsDesired = Math.min(numberOfAreas, 100);
  const intervalSpacing = maxArea / barsDesired;
  let data = [];
  let startPt = 0;
  for (let i = 0; i < barsDesired; i++) {
    let next = startPt + intervalSpacing;
    data.push({
      interval: [startPt, next],
      areas: [],
      percentage: 0,
      startPt: 1 * startPt.toFixed(3)
    });
    startPt = next;
  }

  // Determine the percentage to show along the y-axis and the max value

  for (let area of areas) {
    for (let obj of data) {
      const interval = obj.interval;
      if (area >= interval[0] && area <= interval[1]) {
        obj.areas.push(area);
        break;
      }
    }
  }
  
  let maxPercentage = 0;
  
  data.forEach(obj => {
    obj.percentage = (obj.areas.length / numberOfAreas) * 100;
    maxPercentage = Math.max(maxPercentage, obj.percentage);
  });

  
  useEffect(() => {
    // big help from https://www.d3-graph-gallery.com/graph/barplot_animation_start.html 
    // help with labels from https://bl.ocks.org/d3noob/23e42c8f67210ac6c678db2cd07a747e 
    
    const color = "#69b3a2";
    
    // set the dimensions and margins of the graph
    const margin = {top: 10, right: 30, bottom: 90, left: 70};
    const width = 500 - margin.left - margin.right;
    const height = 500 - margin.top - margin.bottom;

    document.getElementById("graph-container").innerHTML = '';

  // append the svg object to the body of the page
  const svg = d3.select("#graph-container")
    .append("svg")
      .attr("viewBox", `0 0 ${width + margin.left + margin.right} ${height + margin.top + margin.bottom}`)
      .attr("preserveAspectRatio", "xMinYMin meet")
    .append("g")
      .attr("transform",
            "translate(" + margin.left + "," + margin.top + ")");
  
    // X axis
    const x = d3.scaleLinear()
      .range([ 0, width ])
      .domain([0, maxArea])
    svg.append("g")
      .attr("transform", "translate(0," + height + ")")
      .call(d3.axisBottom(x))
      .selectAll("text")
        .attr("transform", "translate(-10,0)rotate(-45)")
        .style("text-anchor", "end");
    
    // text label for the x axis
    svg.append("text")             
    .attr("transform",
          "translate(" + (width/2) + " ," + 
                          (height + margin.top + 50) + ")")
    .style("text-anchor", "middle")
    .text("Area");
    
    // Add Y axis
    const y = d3.scaleLinear()
      .domain([0, maxPercentage])
      .range([ height, 0]);
    svg.append("g")
      .call(d3.axisLeft(y));
    
    // text label for the y axis
    svg.append("text")
        .attr("transform", "rotate(-90)")
        .attr("y", 0 - margin.left)
        .attr("x",0 - (height / 2))
        .attr("dy", "1em")
        .style("text-anchor", "middle")
        .text("Percentage of cells");
    
    // Bars
    svg.selectAll("mybar")
      .data(data)
      .enter()
      .append("rect")
        .attr("x", function(d) { return x(d.startPt); })
        .attr("width", (width / barsDesired) - 1)
        .attr("fill", color)
        // no bar at the beginning thus:
        .attr("height", function(d) { return height - y(0); }) // always equal to 0
        .attr("y", function(d) { return y(0); })
    
    // Animation
    svg.selectAll("rect")
      .transition()
      .duration(800)
      .attr("y", function(d) { return y(d.percentage); })
      .attr("height", function(d) { return height - y(d.percentage); })
      .delay(function(d,i){return(i*50)})
  });

  return (
    <div id="graph-container"></div>
  )
}

export default Graph;