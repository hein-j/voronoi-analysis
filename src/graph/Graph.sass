@use '../App'
  
.graph-container
  @extend %graphic-container
  
  svg
    max-width: 100%
    max-height: 100%
    font-size: 14px
