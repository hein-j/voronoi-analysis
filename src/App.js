import './App.sass';
import Footer from './footer/Footer';
import {useState, useEffect} from 'react';
import getInitialPositions from './helpers/getInitialPositions';
import getApices from './helpers/getApices';
import {Delaunay} from 'd3';
import findFinite from './helpers/findFinite';
import getAreaInfo from './helpers/getAreaInfo';
import findSkewnessAndCoefficient from './helpers/findSkewnessAndCoefficient';
import Graph from './graph/Graph';
import Diagram from './diagram/Diagram';
import Calculations from './calculations/Calculations';
import Popup from './popup/Popup';
import { e } from 'mathjs';

function App() {
  
  function processPositions(positions) {
    const apices = getApices(positions);
    // initiate voronoi
    const buffer = 100;
    const voronoi = Delaunay.from(positions).voronoi([apices.min[0] - buffer, apices.min[1] - buffer, apices.max[0] + buffer, apices.max[1] + buffer]);
    const isFinite = findFinite(voronoi);
    const areasInfo = getAreaInfo(voronoi, isFinite);
    const {skewness, coefficient} = findSkewnessAndCoefficient(areasInfo, voronoi, isFinite);
    return {
      positions: positions,
      apices: apices,
      areas: areasInfo.areas,
      skewness: skewness,
      coefficient: coefficient
    }
  }

  function changePositions(positions) {
    const newObj = processPositions(positions);
    setPositionsObj(newObj);
  }
  
  const [positionsObj, setPositionsObj] = useState(processPositions(getInitialPositions()));

  const [popupObj, setPopupObj] = useState({
    isOpen:false,
    child: <div></div>
  })

  function openPopup (child) {
    setPopupObj({
      isOpen: true,
      child: child
    })
  }

  function closePopup () {
    setPopupObj({
      isOpen: false,
      child: <div></div>
    })
  }


  useEffect(() => {
    window.addEventListener("keydown", (e) => {
      if (e.key !== 'Escape') {
        return;
      }
      closePopup();
    })
  }, [])

  const DiagramComp = <Diagram positions={positionsObj.positions} apices={positionsObj.apices}/>;
  const SkewnessComp = <Calculations name={"Skewness"} value={positionsObj.skewness} />;
  const CoefficientComp = <Calculations name={"Coefficient"} value={positionsObj.coefficient} />;
  const GraphComp = <Graph areas={positionsObj.areas} />;

  return (
    <div className="container">
      <Popup obj={popupObj} close={closePopup} />
      <div className="centered-container">
        <div className="cell" onClick={() => openPopup(DiagramComp)}>
          {DiagramComp}
        </div>
        <div className="cell" onClick={() => openPopup(SkewnessComp)}>
          {SkewnessComp}
        </div>
        <div className="cell" onClick={() => openPopup(CoefficientComp)}>
          {CoefficientComp}
        </div>
        <div className="cell" onClick={() => openPopup(GraphComp)}>
          {GraphComp}
        </div>
      </div>
      <Footer positionsCallBack={changePositions} openPopup={openPopup} closePopup={closePopup}/>
    </div>
  );
}

export default App;
