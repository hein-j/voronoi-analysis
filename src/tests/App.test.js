import { render, screen } from '@testing-library/react';
import App from '../App';
import userEvent from '@testing-library/user-event';
import positionsObj from './positionsObj';
import renderer from 'react-test-renderer';
import Diagram from '../diagram/Diagram';
import Graph from '../graph/Graph';
import Calculations from '../calculations/Calculations';

jest.mock('../downloads/Downloads', () => {return function DummyDownloads(props) {
  return (
    <div data-testid="downloads" />
  );
}})

describe('App', () => {
  test('mounts with correct components', () => {
    render(<App />);
    
    screen.getByTitle('histogram');
    
    screen.getByTitle('diagram');
    
    screen.getByText('Skewness');
    
    screen.getByText('Coefficient');
    
    const calculations = screen.getAllByText((content, element) => {
      return element.tagName.toLowerCase() === 'span' && content.includes('.')
    });
    
    expect(calculations.length).toBe(2);
  });
  
  test('visualizations yield correct pop ups', () => {
    render(<App />);
    userEvent.click(screen.getByTitle('histogram'));
    expect(screen.getAllByTitle('histogram').length).toBe(2);

    userEvent.click(screen.getByTitle('diagram'));
    expect(screen.getAllByTitle('diagram').length).toBe(2);

    userEvent.click(screen.getByText('Skewness'));
    expect(screen.getAllByText('Skewness').length).toBe(2);

    userEvent.click(screen.getByText('Coefficient'));
    expect(screen.getAllByText('Coefficient').length).toBe(2);
  });

  test('footer buttons yield correct pop ups', () => {
    render(<App />);

    userEvent.click(screen.getByRole('button', {
      name: 'download'
    }))

    screen.getByTestId('downloads');

    userEvent.click(screen.getByRole('button', {
      name: 'help'
    }));

    screen.getByText('Usage');
  });

  test('visualizations change correctly on example upload', () => {

    render(<App />);

    const diagram = renderer
      .create(<Diagram positions={positionsObj.positions} apices={positionsObj.apices}/>)
      .toJSON();
      expect(diagram).toMatchSnapshot();

    const graph = renderer
      .create(<Graph areas={positionsObj.areas} />)
      .toJSON();
      expect(graph).toMatchSnapshot();

    const skewness = renderer
      .create(<Calculations name={"Skewness"} value={positionsObj.skewness} />)
      .toJSON();
      expect(skewness).toMatchSnapshot();
    
    const coefficient = renderer
      .create(<Calculations name={"Coefficient"} value={positionsObj.coefficient} />)
      .toJSON();
      expect(coefficient).toMatchSnapshot();
  });

});