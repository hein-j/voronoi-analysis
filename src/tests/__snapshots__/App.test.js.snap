// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App visualizations change correctly on example upload 1`] = `
<div
  className="diagram-container"
/>
`;

exports[`App visualizations change correctly on example upload 2`] = `
<div
  className="graph-container"
/>
`;

exports[`App visualizations change correctly on example upload 3`] = `
<div
  className="calculations-container"
>
  <span
    className="value-name"
  >
    Skewness
  </span>
  <span
    className="value"
  />
</div>
`;

exports[`App visualizations change correctly on example upload 4`] = `
<div
  className="calculations-container"
>
  <span
    className="value-name"
  >
    Coefficient
  </span>
  <span
    className="value"
  />
</div>
`;
